libs/filamat/src/shaders/CodeGenerator.cpp

```diff
@@ -215,6 +215,30 @@ io::sstream& CodeGenerator::generateShaderInputs(io::sstream& out, ShaderType ty
     return out;
 }
 
+utils::io::sstream& CodeGenerator::generateOutput(utils::io::sstream& out, ShaderType type,
+        const utils::CString& name, size_t index,
+        MaterialBuilder::VariableQualifier qualifier,
+        MaterialBuilder::OutputType outputType) const {
+    if (name.empty() || type == ShaderType::VERTEX) {
+        return out;
+    }
+
+    // TODO: add and support additional variable qualifiers
+    (void) qualifier;
+    assert(qualifier == MaterialBuilder::VariableQualifier::OUT);
+
+    const char* typeString = getOutputTypeName(outputType);
+
+    out << "\n#define OUTPUT" << index << " " << name.c_str() << "\n";
+    out << "\n#define OUTPUT_AT" << index << " output_" << name.c_str() << "\n";
+    out << "\n#define OUTPUT_TYPE" << index << " " << typeString << "\n";
+    out << "LAYOUT_LOCATION(" << index << ") out " << typeString <<
```

Comment:
Maybe rename those to "FRAGMENT_OUTPUT_" or "FRAG_OUTPUT_,  "OUTPUT" seems too generic to me, for a #define.